### ISSUE TITLE
Remote clusters installed with istioctl appear to be incompatible in AWS when you use internal load balancers for remote Pilot

### ISSUE THREAD

**Issue by taylorreece** [istio/istio#19639]
**Bug description**

I followed https://istio.io/docs/setup/install/multicluster/shared-vpn/ to install the Istio control plane in one Kubernetes cluster via istioctl, and the "remote" profile istio in a "remote" cluster that is in the same VPC.  Both clusters are EKS clusters in AWS.  Because Pilot is in the main cluster, and the Pilot pod may scale away out of existance, I use an internal load balancer to sit in front of Pilot, so the remote cluster can reference Pilot, as recommended here https://istio.io/docs/setup/install/multicluster/shared-vpn/#deployment-considerations .

This worked fine when I deployed Istio with "helm template".  I would pass in the AWS load balancer endpoint with `--set global.remotePilotAddress=internal-abc-123.us-east-1.elb.amazonaws.com` or whatever, and config maps would properly yield `discoveryAddress: internal-abc-123.us-east-1.elb.amazonaws.com:15011`.  However, it appears that istioctl hard-codes `discoveryAddress: istio-pilot.istio-system:15011`, regardless of if `values.global.remotePilotCreateSvcEndpoint=true` or `false`, and regardless of if you pass in a `--set values.global.remotePilotAddress` argument.

Now, that's fine.  I can set `values.global.remotePilotCreateSvcEndpoint=true` for GCP stacks, as the internal load balancer gets an IP address (as opposed to a FQDN).  However, because AWS uses FQDN's, when istioctl tries to create a local Pilot service that points to a remote location, it errors out saying:

```
Component Base install returned the following errors:
=====================================================
Error: error running kubectl: exit status 1

Error detail:


The Endpoints "istio-pilot" is invalid:  (repeated 1 times)
* subsets[0].addresses[0].ip: Invalid value: "internal-ab47790291d0f11eaa068028d79a51a3-1059610236.us-east-1.elb.amazonaws.com": must be a valid IP address, (e.g. 10.9.8.7) (repeated 1 times)
* subsets[0].addresses[0].ip: Invalid value: "internal-ab47790291d0f11eaa068028d79a51a3-1059610236.us-east-1.elb.amazonaws.com": must be a valid IP address (repeated 1 times)
```
To combat that for now I'm setting 
```
--set values.global.remotePilotAddress=internal-ab47790291d0f11eaa068028d79a51a3-1059610236.us-east-1.elb.amazonaws.com
--set values.global.remotePilotCreateSvcEndpoint=false
```
and then applying a hackish
```
apiVersion: v1
kind: Service
metadata:
    name: istio-pilot
    namespace: istio-system
spec:
  type: ExternalName
  externalName: internal-ab47790291d0f11eaa068028d79a51a3-1059610236.us-east-1.elb.amazonaws.com
  sessionAffinity: None
  ports:
  - name: http-xds
    port: 15010
    protocol: TCP
    targetPort: 15010
  - name: https-xds
    port: 15011
    protocol: TCP
    targetPort: 15011
```

That seems like a bad long-term solution, though.  It would be better if we either:

1) Allowed values.global.remotePilotAddress to be FQDN's in istioctl rather than just IPs
2) Caused istioctl to template out the `injector-mesh` config map to template out like the "helm template" config map, resulting in a `discoveryAddress` that matched `values.global.remotePilotAddress` rather than a hard-coded local service.

Thanks for your help!

**Expected behavior**
I would expect `discoveryAddress` to be the LoadBalancer's endpoint, as opposed to the `istio-pilot.istio-system` service that might not exist in the remote cluster.

**Steps to reproduce the bug**
Create a remote cluster via 
```
istioctl manifest apply \
    --set profile=remote \
    --set autoInjection.enabled=true \
    --set values.global.remotePilotCreateSvcEndpoint=false \
    --set values.global.remotePilotAddress=internal-ab47790291d0f11eaa068028d79a51a3-1059610236.us-east-1.elb.amazonaws.com \
    --set gateways.enabled=false \
    --set values.global.mtls.enabled=true \
    --set values.security.selfSigned=false \
    --set values.global.controlPlaneSecurityEnabled=true
```
To generate a cluster that hard-codes `discoveryAddress: istio-pilot.istio-system:15011` as the discovery address, despite no local istio-pilot service being available.  If you set `values.global.remotePilotCreateSvcEndpoint=true` above, you'll get the error listed in the bug description about IPs vs FQDNs.

**Version**:
```
$ istioctl version --remote
client version: 1.4.2
control plane version: 1.4.2
```

**How was Istio installed?**: Via istioctl (see above)

**Environment where bug was observed (cloud vendor, OS, etc)**
AWS

**Comment by btoonk** [comment]
We're facing the same issue.
As a workaround we're using a overlay for the configmap

`---
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  profile: remote
  autoInjection:
    enabled: true
    components:
      injector:
        k8s:
          overlays:
          - kind: ConfigMap
            name: injector-mesh
            patches:
            - path: data.mesh
              value: |-
                # Unix Domain Socket through which envoy communicates with NodeAgent SDS to get
                   # key/cert for mTLS. Use secret-mount files instead of SDS if set to empty.
                   sdsUdsPath: ""
                   defaultConfig:
                     #
                     # TCP connection timeout between Envoy & the application, and between Envoys.
                     connectTimeout: 10s
                     #
                     ### ADVANCED SETTINGS #############
                     # Where should envoy's configuration be stored in the istio-proxy container
                     configPath: "/etc/istio/proxy"
                     # The pseudo service name used for Envoy.
                     serviceCluster: istio-proxy
                     # These settings that determine how long an old Envoy
                     # process should be kept alive after an occasional reload.
                     drainDuration: 45s
                     parentShutdownDuration: 1m0s
                     #
                     # Port where Envoy listens (on local host) for admin commands
                     # You can exec into the istio-proxy container in a pod and
                     # curl the admin port (curl http://localhost:15000/) to obtain
                     # diagnostic information from Envoy. See
                     # https://lyft.github.io/envoy/docs/operations/admin.html
                     # for more details
                     proxyAdminPort: 15000
                     #
                     # Set concurrency to a specific number to control the number of Proxy worker threads.
                     # If set to 0 (default), then start worker thread for each CPU thread/core.
                     concurrency: 2
                     #
                     tracing:
                       zipkin:
                         # Address of the Zipkin collector
                         address: zipkin.istio-system:9411
                     #
                     # Mutual TLS authentication between sidecars and istio control plane.
                     controlPlaneAuthPolicy: MUTUAL_TLS
                     #
                     # Address where istio Pilot service is running
                     discoveryAddress: <REPLACE WITH PILOT ADDRESS>:15011`

which can be executed with `istioctl manifest generate -f <FILE>`

I think this needs to be fixed in the file https://github.com/istio/installer/blob/master/istio-control/istio-autoinject/templates/configmap.yaml

The file https://github.com/istio/installer/blob/master/istio-control/istio-discovery/templates/configmap.yaml already contains the logic for remotePilotAddress

**Comment by taylorreece** [comment]
Thanks for pointing me in the right direction @btoonk .  I've submitted a PR for istio/installer that'll hopefully trickle its way down into a future istio release. https://github.com/istio/installer/pull/635

**Comment by taylorreece** [comment]
This missed the 1.4.3 release.  🤞 for the 1.4.4 release.

**Comment by arizvi002** [comment]
Hi folks, I was trying to implement cross cluster also and was wondering when 1.4.4 is releasing, does anyone have any idea?

**Comment by taylorreece** [comment]
@arizvi002 no idea on 1.4.4, nor do I know if this'll be cherry-picked into 1.4.*, but 1.5.0 comes out in about a month https://github.com/istio/istio/wiki/Istio-Release-1.5

**Comment by arizvi002** [comment]
@taylorreece Thanks for quick response. 
Im not seeing any mention of your changes on the notes. So no guaranteeing it'll be released.
Were you able to get the workaround @btoonk mentioned working? Or did you use some other method? I'm pretty new to K8s/Istio so any help is appreciated, been stuck on this for a while.

**Comment by taylorreece** [comment]
@arizvi002 I have a really crappy work-around where I grab the first internal IP address of the LB, and use that as my remote pilot / telemetry / policy endpoint.
```
lb_endpoint=$(kubectl get svc $lb_name -n istio-system --context ${INGRESS_CLUSTER_CONTEXT} --template "{{ range .status.loadBalancer.ingress}}{{.hostname}}{{end}}")
lb_ip=$(dig +short $lb_endpoint | head -n 1)
```

**Comment by arizvi002** [comment]
Thanks @taylorreece 
This is what I was doing too using "host <elb cname>" I get 

kubectl exec --context=$CTX_CLUSTER1 -it -n sample -c sleep $(kubectl get pod --context=$CTX_CLUSTER1 -n sample -l app=sleep -o jsonpath='{.items[0].metadata.name}') -- curl helloworld.sample:5000/hello
Hello version: v1, instance: helloworld-v1-6757db4ff5-2r67r

kubectl exec --context=$CTX_CLUSTER2 -it -n sample -c sleep $(kubectl get pod --context=$CTX_CLUSTER2 -n sample -l app=sleep -o jsonpath='{.items[0].metadata.name}') -- curl helloworld.sample:5000/hello
Hello version: v1, instance: helloworld-v1-6757db4ff5-2r67r

but I also get 

kubectl exec --context=$CTX_CLUSTER1 -it -n sample -c sleep $(kubectl get pod --context=$CTX_CLUSTER1 -n sample -l app=sleep -o jsonpath='{.items[0].metadata.name}') -- curl helloworld.sample:5000/hello
curl: (6) Could not resolve host: helloworld.sample
command terminated with exit code 6

Are you doing anything to handle this?

**Comment by taylorreece** [comment]
@arizvi002 in case you're curious, this didn't make the 1.4.4 release.  Hopefully it'll be in the 1.5.0... I'm waiting for 1.5.0-beta.1 to be compiled and available to test; it wasn't in 1.5.0-alpha.0.

It looks like someone cherry-picked the patch into the 1.4 branch 4 hours ago, so maybe it'll be in 1.4.5, if that ever exists.
